bin/bootstrap_practice_exercise.exs

```diff
@@ -0,0 +1,283 @@
+            [{"input", nil}]
+          end,
+        error: match?(%{"error" => _err}, expected)
+      }
+    }
+
+  def explore_properties(_data), do: %{}
+
+  def print_property(%{name: name, variables: variables, error: error}) do
+    return_type = if error, do: "{:ok, TODO.t()} | {:error, String.t()}", else: "TODO.t()"
+
+    variable_types =
+      Enum.map_join(variables, ", ", fn
+        {var, nil} ->
+          "#{var} :: TODO.t()"
+
+        {var, sub_vars} ->
+          "#{var} :: %{#{Enum.map_join(sub_vars, ", ", &(&1 <> " : TODO.t()"))}}"
+      end)
+
+    variable_list = Enum.map_join(variables, ", ", fn {var, _} -> var end)
+
+    """
+     @doc \"\"\"
+     TODO: add function description and replace types in @spec
+     \"\"\"
+     @spec #{name}(#{variable_types}) :: #{return_type}
+     def #{name}(#{variable_list}) do
+     end
+    """
+  end
+
+  def print_comments(comments, do_not_print) do
+    comments
+    |> Enum.reject(fn {field, _value} -> field in do_not_print end)
+    |> Enum.map_join("\n", fn
+      {"comments", values} when is_list(values) ->
+        "# #{Enum.map_join(values, "\n# ", &String.trim/1)}"
+
+      {field, values} when is_list(values) ->
+        "#\n# --#{field} --\n# #{Enum.map_join(values, "\n# ", &inspect/1)}"
+
+      {field, value} ->
+        "#\n# -- #{field} --\n# #{inspect(value)}"
+    end)
+  end
+
+  def print_input(%{} = input),
+    do: Enum.map_join(input, "\n", fn {variable, value} -> "#{variable} = #{inspect(value)}" end)
+
+  def print_input(input), do: "input = #{inspect(input)}"
+
+  def print_expected(%{"error" => err}, _error), do: "{:error, #{inspect(err)}}"
+  def print_expected(expected, true), do: "{:ok, #{inspect(expected)}}"
+  def print_expected(expected, false), do: inspect(expected)
+
+  def print_test_case(
+        %{"description" => description, "cases" => sub_cases} = category,
+        properties,
+        module
+      ) do
+    """
+    describe \"#{description}\" do
+    #{Generate.print_comments(category, ["description", "cases"])}
+    #{Enum.map_join(sub_cases, "\n\n", &Generate.print_test_case(&1, properties, module))}
+    end
+    """
+  end
+
+  def print_test_case(
+        %{
+          "description" => description,
+          "property" => property,
+          "input" => input,
+          "expected" => expected
+        } = test,
+        properties,
+        module
+      ) do
+    %{name: name, variables: variables, error: error} = properties[property]
+    variable_list = Enum.map_join(variables, ", ", fn {var, _} -> var end)
+
+    """
+    @tag :pending
+    test \"#{description}\" do
+      #{Generate.print_comments(test, ["description", "property", "input", "expected", "uuid"])}
+      #{print_input(input)}
+      output = #{module}.#{name}(#{variable_list})
+      expected = #{print_expected(expected, error)}
+
+      assert output == expected
+    end
+    """
+  end
+end
+
+Mix.install([
+  {:jason, "~> 1.2"},
+  {:toml, "~> 0.6"}
+])
```

Comment:
I don't know if there is really a standard established for mix install, but can we hoist this to the top of the file?